include/EquationSystem.h

```diff
@@ -353,6 +353,7 @@ class EquationSystem
   std::string dofName_{"undefined"};
 
   bool extractDiagonal_{false};
+  ScalarFieldType* Udiag_{nullptr};
```

Comment:
Could this be a method instead? something like `get_diagonal_field()` that returns `nullptr` in base implementation and is specialized for `MomentumEquationSystem` where it returns `Udiag_`?

Reply:
Yes that's a good idea, I'll do that.